site/content/3.12/data-science/graph-analytics.md

```diff
@@ -0,0 +1,667 @@
+
+*Label Propagation* can be used to implement community detection on large
+graphs. The algorithm assigns a community, more precisely, a Community ID 
+(a natural number), to every vertex in the graph. 
+The idea is that each vertex should be in the community that most of
+its neighbors are in. 
+
+At first, the algorithm assigns unique initial Community IDs to the vertices. 
+The assignment is deterministic given the graph and the distribution of vertices
+on the shards, but there is no guarantee that a vertex obtains
+the same initial ID in two different runs of the algorithm, even if the graph does not change 
+(because the sharding may change). Moreover, there is no guarantee on a particular
+distribution of the initial IDs over the vertices.
+
+Then, in each iteration, a vertex sends its current Community
+ID to all its neighbor vertices. After that each vertex adopts the Community ID it
+received most frequently in the last step. 
+
+Note that, in a usual implementation of Label Propagation, if there are
+multiple most frequently received Community IDs, one is chosen randomly.
+An advantage of our implementation is that this choice is deterministic.
+This comes for the price that the choice rules are somewhat involved: 
+If a vertex obtains only one ID and the ID of the vertex from the previous step,
+its old ID, is less than the obtained ID, the old ID is kept. 
+(IDs are numbers and thus comparable to each other.) If a vertex obtains
+more than one ID, its new ID is the lowest ID among the most frequently 
+obtained IDs. (For example, if the obtained IDs are 1, 2, 2, 3, 3,
+then 2 is the new ID.) If, however, no ID arrives more than once, the new ID is
+the minimum of the lowest obtained IDs and the old ID. (For example, if the
+old ID is 5 and the obtained IDs are 3, 4, 6, then the new ID is 3. 
+If the old ID is 2, it is kept.) 
+
+If a vertex keeps its ID 20 times or more in a row, it does not send its ID.
+Vertices that did not obtain any IDs do not update their ID and do not send it.
+
+The algorithm runs until it converges, which likely never really happens on
+large graphs. Therefore you need to specify a maximum iteration bound.
+The default bound is 500 iterations, which is too large for
+common applications. 
+
+The algorithm should work best on undirected graphs. On directed
+graphs, the resulting partition into communities might change, if the number 
+of performed steps changes. How strong the dependence is
+may be influenced by the density of the graph.
+
+Parameters:
+- `graph_id`
+- `start_label_attribute`
+- `synchronous`
+- `random_tiebreak`
+- `maximum_supersteps`
+
+Result: a community ID for each vertex
+
+```bash
+GRAPH_ID="234"
+curl -H "Authorization: bearer $ADB_TOKEN" -XPOST -d "{\"graph_id\":$GRAPH_ID,\"start_label_attribute\":\"start_attr\",\"synchronous\":false,\"random_tiebreak\":false,\"maximum_supersteps\":500}" "$ENGINE_URL/v1/labelpropagation"
+```
+
+##### Speaker-Listener Label Propagation
```

Comment:
I do not think this is exactly what we have implemented as "attribute propagation".

Reply:
If I understand correctly, SLPA only propagates a subset of the labels based on frequency whereas attribute propagation propagates all of them?